### ISSUE TITLE
Wrong arguments for variadic function

### ISSUE THREAD

**Issue by belka-ew** [ldc-developers/ldc#2613]
The code is a bit long, it depends on `undead`'s socket streams, but I'll point out where the problem comes from.

ServerTest.d:
```d
import core.thread;
import std.conv;
import std.socket;
import std.stdio;
import undead.socketstream;
import undead.stream;

void write2out(OutputStream output)
{
    output.writef("%s %s\r\n", "HTTP/1.1", "OK");
    output.writeString("\r\n");
    output.flush;
}

void main()
{
    Socket server = new TcpSocket;
    server.setOption(SocketOptionLevel.SOCKET, SocketOption.REUSEADDR, true);
    server.bind(new InternetAddress("localhost", InternetAddress.PORT_ANY));

    InternetAddress address = new InternetAddress("localhost", (cast(std.socket.InternetAddress) server.localAddress).port);

    server.listen(32);

    Socket client = new TcpSocket;

    void run() /** Start */
    {
        while (true)
        {
            try
            {
                auto socket = server.accept;

                scope (exit)
                {
                    socket.shutdown(SocketShutdown.BOTH);
                    socket.close;
                }

                auto output = new SocketStream(socket, FileMode.Out);

                write2out(output);
            }
            catch (SocketAcceptException)
            {
                continue;
            }
            return;
        }
    }

    new Thread(&run).start;
    Thread.sleep(100.msecs);
    /** Start */

    client.connect(address);

    auto output = new SocketStream(client);

    // when
    output.writeString("POST / HTTP/1.1\r\nContent-Length: 9\r\n\r\n<PACKET/>");
    output.flush;

    auto input = new SocketStream(client);
    char[50] line;
    size_t i = 0;
    char c;

    while (c != '\r')
    {

        input.read(c);

        line[i] = c;
        ++i;
    }
    writeln(line[0 .. i]);
}
```
then:
```shell
git clone https://github.com/dlang/undeaD.git
cd undeaD
git checkout 14540c73f8116fe0dd6e98fdedc03e3af2678c74

dub build --compiler=../ldc2-1.8.0-linux-x86_64/bin/ldc2
cd ..
./ldc2-1.8.0-linux-x86_64/bin/ldc2 -I undeaD/src undeaD/bin/libundead.a ServerTest.d
./ServerTest
```
Output:
> HTTP/1.1OK

The same procedure with dmd will result in (expected):
> HTTP/1.1 OK

The root of the bug is somewhere around `write2out`. Look at `output.writef("%s %s\r\n", "HTTP/1.1", "OK");`. The first argument is ignored, no matter what it contains. The remaining arguments are written to the stream as `output.writeln` would do. So wrong function with wrong arguments seems to be called.

**Comment by belka-ew** [comment]
```
./ldc2 --version

LDC - the LLVM D compiler (1.8.0):
  based on DMD v2.078.3 and LLVM 5.0.1
  built with LDC - the LLVM D compiler (1.8.0)
  Default target: x86_64-unknown-linux-gnu
  Host CPU: skylake
  http://dlang.org - http://wiki.dlang.org/LDC

  Registered Targets:
    aarch64    - AArch64 (little endian)
    aarch64_be - AArch64 (big endian)
    arm        - ARM
    arm64      - ARM64 (little endian)
    armeb      - ARM (big endian)
    nvptx      - NVIDIA PTX 32-bit
    nvptx64    - NVIDIA PTX 64-bit
    ppc32      - PowerPC 32
    ppc64      - PowerPC 64
    ppc64le    - PowerPC 64 LE
    thumb      - Thumb
    thumbeb    - Thumb (big endian)
    x86        - 32-bit X86: Pentium-Pro and above
    x86-64     - 64-bit X86: EM64T and AMD64
```

**Comment by kinke** [comment]
Thanks for the report. I had a quick look into undead and its formatting code, but gave up as it's too time consuming. It'd be a great help if you could do some further investigations to try to nail this further down, maybe with some printf outputs in the formatting functions etc.

**Comment by belka-ew** [comment]
```d
import core.vararg;
import std.stdio;

interface OutputStream
{
    void writef(int i, ...);
}


class Stream : OutputStream
{
    void writef(int i, ...)
    {
        va_list list;
        va_start(list, i);

        writeln(va_arg!string(list));

        va_end(list);
    }
}

void write2out(OutputStream output)
{
    output.writef(0, "%s %s\r\n");
}

void main()
{
    write2out(new Stream);
}
```
Expected output:
> %s %s
>

Actual output is an empty line.

**Comment by kinke** [comment]
Thanks, that definitely helped. It works when changing the param type of `void write2out(OutputStream output)` to `Stream`; the interface thunk apparently doesn't forward any variadics:
```LL
define linkonce_odr void @_DThn16_7current6Stream6writefMFiYv(%current.Stream* nonnull %.this_arg, { i64, %object.TypeInfo** } %._arguments, i32 %i_arg, ...) unnamed_addr #1 comdat {
  %1 = bitcast %current.Stream* %.this_arg to i8* ; [#uses = 1]
  %2 = getelementptr inbounds i8, i8* %1, i32 -16 ; [#uses = 1, type = i8*]
  %3 = bitcast i8* %2 to %current.Stream*         ; [#uses = 1]
  tail call void (%current.Stream*, { i64, %object.TypeInfo** }, i32, ...) @_D7current6Stream6writefMFiYv(%current.Stream* %3, { i64, %object.TypeInfo** } %._arguments, i32 %i_arg)
  ret void
}
```

**Comment by kinke** [comment]
Fixed by #2630.
